### PR TITLE
Added ability to rebind control key

### PR DESCRIPTION
**Pull request by mzegar**
Implements #12658

runelite_after

### CODE REVIEW

**Comment by Adam-** [comment]
Did you look at why it is throwing an exception? That might mean something else is wrong.

**Comment by mzegar** [comment]
> Did you look at why it is throwing an exception? That might mean something else is wrong.

Yeah, I am just completely lost as to why the exception is being thrown. I attempted to make break points in the `.matches(e)` method, but the exception is thrown right after that method is called.

I'm assuming it has something to do with the fact that `mappedKeyCode` cannot already be assigned.

Let me know if there is any other helpful information I could provide.

**Comment by Adam-** [comment]
The stack trace should tell you why... can you push a branch that the test fails on? I could look.

**Comment by mzegar** [comment]
> The stack trace should tell you why... can you push a branch that the test fails on? I could look.

I just pushed it directly to the branch. Check out `KeyRemappingListenerTest.java`. I appreciate the help, I genuinely gave it 30 minutes or so but was just extremely dumbfounded. I use a debugger all the time and I've never seen such behavior (maybe because it's in a test?)

**Comment by Adam-** [comment]
This is just because `config.control()` is `null` in the test due to it not being mocked.

**Comment by mzegar** [comment]
> This is just because `config.control()` is `null` in the test due to it not being mocked.

Thanks, I going to look into how I missed that. Will fix, thanks for taking a look at it.

**Comment by mzegar** [comment]
@Adam- fixed, thanks again for the help.

**Comment by mzegar** [comment]
@Adam- Fixed, thanks for the help.

File: runelite-client/src/main/java/net/runelite/client/plugins/keyremapping/KeyRemappingConfig.java [hunk 1]
```diff
@@ -276,14 +277,34 @@ default ModifierlessKeybind esc()
 		return new ModifierlessKeybind(KeyEvent.VK_ESCAPE, 0);
 	}
 
+	@ConfigSection(
+		name = "Miscellaneous Remapping",
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/keyremapping/KeyRemappingConfig.java:281]: I don't think a new section for this is required. Having no section usually implies miscellaneous.

File: runelite-client/src/test/java/net/runelite/client/plugins/keyremapping/KeyRemappingListenerTest.java [hunk 2]
```diff
@@ -73,6 +74,7 @@ public void testTypingStateChange()
 		when(keyRemappingConfig.down()).thenReturn(new ModifierlessKeybind(KeyEvent.VK_S, 0));
 		when(keyRemappingConfig.left()).thenReturn(new ModifierlessKeybind(KeyEvent.VK_A, 0));
 		when(keyRemappingConfig.right()).thenReturn(new ModifierlessKeybind(KeyEvent.VK_D, 0));
+		when(keyRemappingConfig.control()).thenReturn(new ModifierlessKeybind(KeyEvent.VK_CONTROL, InputEvent.CTRL_DOWN_MASK));
```
Comment (Adam-) [inline runelite-client/src/test/java/net/runelite/client/plugins/keyremapping/KeyRemappingListenerTest.java:77]: You can also just put this in `setUp` I think

File: runelite-client/src/main/java/net/runelite/client/plugins/keyremapping/KeyRemappingConfig.java [hunk 3]
```diff
@@ -286,4 +287,15 @@ default ModifierlessKeybind space()
...
+
+	@ConfigItem(
+		position = 22,
+		keyName = "control",
+		name = "Control",
+		description = "The key which will replace {Control}."
+	)
+	default ModifierlessKeybind control()
+	{
+		return new ModifierlessKeybind(KeyEvent.VK_CONTROL, InputEvent.CTRL_DOWN_MASK);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/keyremapping/KeyRemappingConfig.java:299]: From testing on my system and talking to abex, it seems like this should be `new ModifierlessKeybind(KeyEvent.VK_UNDEFINED, InputEvent.CTRL_DOWN_MASK);`. Optionally we could just make this not have a default value and check for null, although we don't do that for the other keybinds.

**Pull request merged**